contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,739 @@
+Whereas `kubectl exec` runs a *process* in a *container*, `kubectl debug` will
+be similar but run a *container* in a *pod*.
+
+A container created by `kubectl debug` is a *Debug Container*. Just like a
+process run by `kubectl exec`, a Debug Container is not part of the pod spec and
+has no resource stored in the API. Unlike `kubectl exec`, a Debug Container
+*does* have status that is reported in `v1.PodStatus` and displayed by `kubectl
+describe pod`.
+
+For example, the following command would attach to a newly created container in
+a pod:
+
+```
+kubectl debug -c debug-shell --image=debian target-pod -- bash
+```
+
+It would be reasonable for Kubernetes to provide a default container name and
+image, making the minimal possible debug command:
+
+```
+kubectl debug target-pod
+```
+
+This creates an interactive shell in a pod which can examine and signal other
+processes in the pod. It has access to the same network and IPC as processes in
+the pod. It can access the filesystem of other processes by `/proc/$PID/root`.
+As is already the case with regular containers, Debug Containers can enter
+arbitrary namespaces of another container via `nsenter` when run with
+`CAP_SYS_ADMIN`.
+
+*Please see the User Stories section for additional examples and Alternatives
+Considered for the considerable list of other solutions we considered.*
+
+## Implementation Details
+
+The implementation of `kubectl debug` closely mirrors the implementation of
+`kubectl exec`, with most of the complexity implemented in the `kubelet`. How
+functionality like this best fits into Kubernetes API has been contentious. In
+order to make progress, we will start with the smallest possible API change,
+extending `/exec` to support Debug Containers, and iterate.
+
+From the perspective of the user, there's a new command, `kubectl debug`, that
+creates a Debug Container and attaches to its console. We believe a new command
+will be less confusing for users than overloading `kubectl exec` with a new
+concept. Users give Debug Containers a name (e.g. "debug" or "shell") which can
+subsequently be used to reattach and is reported by `kubectl describe`.
+
+### Kubernetes API Changes
+
+#### Chosen Solution: "exec++"
+
+We will extend `v1.Pod`'s `/exec` subresource to support "executing" container
+images. The current `/exec` endpoint must implement `GET` to support streaming
+for all clients. We don't want to encode a (potentially large) `v1.Container` as
+an HTTP parameter, so we must extend `v1.PodExecOptions` with the specific
+fields required for creating a Debug Container:
+
+```
+// PodExecOptions is the query options to a Pod's remote exec call
+type PodExecOptions struct {
+        ...
+        // DebugName is the name of the Debug Container. Its presence will cause
+        // exec to create a Debug Container rather than performing a runtime exec.
+        DebugName string `json:"debugName,omitempty" ...`
```

Comment:
I vote for making a sub section:

```go 
type PodExecOptions struct {
  ...
  EphemeralContainer *PodExecEphemeralContainerSpec
}

type PodExecEphemeralContainerSpec struct {
  Name string
  Image string
}
```

Reply:
SGTM, though in [my naive prototype](https://github.com/verb/kubernetes/commit/c6a72efad4c72b2f7815c07d295b7ac9141199f5#diff-c73f80ad83608f18657d22a06950d929R475) the sub section wasn't populated from the HTTP params. 

I'm not very familiar with the api machinery so I probably just missed something. I see where [queryparams flattens the struct based on JSON names](https://github.com/kubernetes/apimachinery/blob/master/pkg/conversion/queryparams/convert.go#L129). Do I need to write a custom converter somewhere to get it back from params to an object? Could you point me in the right direction?

Reply:
yeah, nested structs parsed from query params won't work cleanly (https://github.com/kubernetes/kubernetes/issues/21476)

Reply:
Earlier in the text it said this was not using query params--I thought it was using a message sent after the SPDY channel was opened.